prereise/cli/data_sources.py

```diff
@@ -0,0 +1,130 @@
+    @property
+    @abstractmethod
+    def command_help(self):
+        """String with a help description for the command. See add_parser function
+        at `this link <https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_subparsers>`_
+        for more details
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def extract_arguments(self):
+        """List of dictionaries defining the various flags and parameters
+        the user needs to pass into this command, which are
+        subsequently passed to the below extract function.
+        See `add_argument <https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_argument>`_
+        to understand what is acceptable/unacceptable in these
+        dictionaries.
+        """
+        pass
+
+    @abstractmethod
+    def extract(self):
+        """Function that takes in the arguments defined in
+        the above property extract_arguments and downloads
+        the appropriate data and saves it somewhere.
+        """
+        pass
+
+
+class WindData(DataSource):
+    @property
+    def command_name(self):
+        """See :py:func:`prereise.cli.data_sources.DataSource.command_name`
+
+        :return: (*str*)
+        """
+        return "wind_data"
+
+    @property
+    def command_help(self):
+        """See :py:func:`prereise.cli.data_sources.DataSource.command_help`
+
+        :return: (*str*)
+        """
+        return "Download wind data"
+
+    @property
+    def extract_arguments(self):
+        """See :py:func:`prereise.cli.data_sources.DataSource.extract_arguments`
+
+        :return: (*str*)
+        """
+        return [
+            {
+                "command_flags": ["--region", "-r"],
+                "required": True,
+                "choices": ["Texas", "Eastern", "Western"],
+                "type": str,
+                "action": "append",
+            },
+            {
+                "command_flags": ["--start_date", "-sd"],
+                "required": True,
+                "type": validate_date,
+            },
+            {
+                "command_flags": ["--end_date", "-ed"],
+                "required": True,
+                "type": validate_date,
+            },
+            {
+                "command_flags": ["--file_path", "-fp"],
+                "required": True,
+                "type": validate_file_path,
+            },
+        ]
+
+    def extract(self, region, start_date, end_date, file_path, **kwargs):
+        """See :py:func:`prereise.cli.data_sources.DataSource.extract`
+
+        :param list region: list of regions to download wind farm data for
+        :param str start_date: date designating when to start the data download
+        :param str end_date: date designating when to end the data download
+        :param str file_path: file location on local filesystem on where to store the data
+        """
+        regions = list(set(region))
+        assert start_date <= end_date
```

Comment:
Since the `strptime` format check doesn't require a leading zero, this has the potential to produce some weird errors where `-sd 2021-5-1 -ed 2021-12-1` fails this assertion even though they are both valid dates and the start date is before the end date.

Reply:
great catch, updated to use strptime at this assertion point as well